src/System.Linq/src/System/Linq/Where.cs

```diff
@@ -271,34 +432,86 @@ public override Iterator<TResult> Clone()
                 return new WhereSelectArrayIterator<TSource, TResult>(_source, _predicate, _selector);
             }
 
-            public override bool MoveNext()
+            public int GetCount(bool onlyIfCheap)
             {
-                if (_state == 1)
+                if (onlyIfCheap)
                 {
-                    while (_index < _source.Length)
+                    return -1;
+                }
+
+                int count = 0;
+
+                foreach (TSource item in _source)
+                {
+                    if (_predicate(item))
                     {
-                        TSource item = _source[_index];
-                        _index++;
-                        if (_predicate(item))
+                        checked
                         {
-                            _current = _selector(item);
-                            return true;
+                            count++;
```

Comment:
I think we must run the selectors. It is reasonable for the user that selectors will run. 
In fact, it does not seem very unlikely that user may do ".Count" specifically just to run sideeffecting selectors.


---
In reply to: [88807476](https://github.com/dotnet/corefx/pull/12703#discussion_r88807476) [](ancestors = 88807476)